### DISCUSSION TITLE
How to modify values of many to many relationship in the crud interface

### DISCUSSION THREAD

**Discussion by dices** [fastapi/fastapi#9115] [Questions]
### First check

* [X] I added a very descriptive title to this issue.
* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [X] I already read and followed all the tutorial in the docs and didn't find an answer.
* [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [X] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

Hello everyone. I am having a really hard time updating a user with a many to many relationship.
The models used are these.

```Python
class UserBase(BaseModel):
    email: Optional[EmailStr] = None
    is_active: Optional[bool] = True
    is_superuser: bool = False
    full_name: Optional[str] = None
    roles: List[Role] = []

class User(Base):
    id = Column(Integer, primary_key=True, index=True)
    full_name = Column(String, index=True)
    email = Column(String, unique=True, index=True, nullable=False)
    hashed_password = Column(String, nullable=False)
    is_active = Column(Boolean(), default=True)
    is_superuser = Column(Boolean(), default=False)
    items = relationship("Item", back_populates="owner")
    roles = relationship(
        'Role', secondary=user_role,
        backref=backref("users")
    )
```
All the models have Config orm=True

The usage is that I have an admin panel where I can add roles to a profile. So I get the json model to the Vue app, add new roles and then post it to the api. When this updated model tries to set the roles attribute it fails.

Now after the endpoint is called. I see the new roles in the pydantic model as expected. But when the crud update tries to apply the new roles in the list it fails.

But when it reaches this point
```Python
    def update(
        self,
        db: Session,
        *,
        db_obj: ModelType,
        obj_in: Union[UpdateSchemaType, Dict[str, Any]]
    ) -> ModelType:
        obj_data = jsonable_encoder(db_obj)
        if isinstance(obj_in, dict):
            update_data = obj_in
        else:
            update_data = obj_in.dict(exclude_unset=True)
        for field in obj_data:
            if field in update_data:
                setattr(db_obj, field, update_data[field]) <------- Fails here
        db.add(db_obj)
        db.commit()
        db.refresh(db_obj)
        return db_obj
```

When I try to update user profile and add more roles to the list. I get the following error.
backend_1 | File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py", line 45, in append
backend_1 | item_state = attributes.instance_state(item)
backend_1 | AttributeError: 'dict' object has no attribute '_sa_instance_state'

Any idea what I can be doing incorrectly ?
Using fastapi = "^0.60.1" and Python 3.8
Thanks

**Comment by ghost** [comment]
A Quick glance at this tells me that it's a SQLAlchemy Error, and not a FastAPI Error or Bug, Perhaps check the SQLAlchemy Documentation for it?

**Comment by dices** [comment]
> A Quick glance at this tells me that it's a SQLAlchemy Error, and not a FastAPI Error or Bug, Perhaps check the SQLAlchemy Documentation for it?

I looked at the documentation, but they create the models and add them one by one in the examples. I am trying to understand the method used in the fastapi project generator with postgres and how to modify it to update nested relationships in the orm.

**Comment by ghost** [comment]
OK, I see the issue here. I think you're misunderstanding what a `relationship` is! The Gist of a relationship is that it is a "Query" That just specifies how the two tables join, it just filters the other records by using the key it is related to. Think of a relationship field as a Query that essentially asks the other table (In this case, Roles) of the records that uses a foreign key (In this case, it Asks the Roles Table about the records that use this particular User_id)

With that out of the way! 

A `relationship` "updates" as you insert things into the Roles Table directly, and similarly, it also changes when you actually update an existing record in that particular table.

Take a look at the dictionary you've obtained from `json_encoder`: 

`{'_sa_instance_state': <sqlalchemy.orm.state.InstanceState object at 0x7f9d04563810>}`

This isn't what you want, obviously! What you can do is to create an instance of the Model that you are trying to update!

We can obtain a Record by Querying it from the database: ```rec = db.query(Roles).filter(Roles.id == id).first()```

Note that the filter used here is completely arbitrary.

Then Update Each of those fields?

```py
for (Key, Value) in obj_in.items():
    setattr(rec, Key, Value) # Note, do not ever actually do this. I'm doing to remain very close to your original code.

# Updated? Flush and Commit the record.
db.commit() 

```

As a side note, this really is a problem with SQLAlchemy (Not a bug, rather the usage), and the lovely, *wonderful* and absolutely soothing documentation over at [SQLAlchemy Official Documentation](https://docs.sqlalchemy.org/en/13/orm/tutorial.html) is a great place to start. It is very through and gives you examples to follow along if you do not understand.  Please follow the Examples from FastAPI Main Documentation as well, they are most helpful in Integrating SQLAlchemy in FastAPI.

**Comment by amirjani** [comment]
Hi, 

after looking at [SQLAlchemy](https://docs.sqlalchemy.org/en/13/orm/tutorial.html#building-a-many-to-many-relationship) I realized that for having many to many relations like ```user_role``` we should do something like below code: 

```python
"""
file: user.py
"""
class User(Base):
    __tablename__ = "user"

    id = Column(Integer, primary_key=True, index=True)
    full_name = Column(String, index=True)
    email = Column(String, unique=True, index=True, nullable=False)
    hashed_password = Column(String, nullable=False)
    is_active = Column(Boolean(), default=True)
    is_superuser = Column(Boolean(), default=False)

    items = relationship("Role", secondary=UserRole.__tablename__, back_populates="user")
```

```python
"""
file: role.py
"""
class Role(Base):
    __tablename__ = 'role'

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, index=True)
    is_primary = Column(Boolean)

    items = relationship("User", secondary=UserRole.__tablename__, back_populates="role")
```

```python
"""
file: user_role.py
"""
class UserRole(Base):
    __tablename__ = 'user_role'

    user_id = Column('user_id', ForeignKey('user.id'), primary_key=True)
    role_id = Column('role_id', ForeignKey('role.id'), primary_key=True)
```


**don't forget to import UserRole class in User and Role classes**

**Comment by Silur** [comment]
the problem is that `jsonable_encoder` returns an empty object `{}` when called on a model that contains a many-to-many relationship, thus you cannot use inherited functions from `crud/base.py`

**Comment by jrujano-zz** [comment]
> the problem is that `jsonable_encoder` returns an empty object `{}` when called on a model that contains a many-to-many relationship, thus you cannot use inherited functions from `crud/base.py`

I have established a many-to-many relationship here in FAST-SQLAlchemy, but not the slightest idea of how to generate the insert and update data. Any suggestion

```` 
# models.py
association_table = Table('permissionsroles', Base.metadata,
    Column('role_id', ForeignKey('role.id'), primary_key=True),
    Column('permissions_id', ForeignKey('permissions.id'), primary_key=True)
)

class Permissions(Base):
   id = Column(Integer, primary_key=True, index=True)
   label = Column(String, nullable=False)
   name = Column(String, unique=True, index=True, nullable=False)
   aplication_id = Column(Integer, ForeignKey("application.id"))
   time_created = Column(DateTime(timezone=True), server_default=func.now())
   time_updated = Column(DateTime(timezone=True), onupdate=func.now())
   app = relationship("Application", back_populates="permissions")

   role = relationship("Role",
                    secondary=association_table,
                    backref="permissions_role")
   
class Role(Base):
   id = Column(Integer, primary_key=True, index=True)
   name = Column(String, unique=True, index=True, nullable=False)
   is_active = Column(Boolean(), default=True)
   time_created = Column(DateTime(timezone=True), server_default=func.now())
   time_updated = Column(DateTime(timezone=True), onupdate=func.now())
```` 
```` 
#schema.py

class PermissionsSchema(BaseModel):
    id: Optional[int]
    name: str
    class Config:
        orm_mode = True
        
class RoleIn(BaseModel):
    name: str
    role: List[PermissionsSchema]

    class Config:
        orm_mode = True
```` 

```` 
#endpoint.py
@router.post("/role", response_model=schemas.RoleIn)
def create_role(*,db: Session = Depends(deps.get_db), role_in: schemas.RoleIn)-> Any:
    role = crud.security.role.get_role_by_name (db, name=role_in.name)
    if role:
        raise HTTPException(
            status_code=400,
            detail="The user with this role  already exists in the system.",
        )
    else:
        role_data = jsonable_encoder(role_in)
        print(role_data)
        db_role = Role(**role_data)
        db.add(db_role)
        db.commit()
        db.refresh(db_role)
    return db_role
```` 

**I have established a many-to-many relationship here in FAST-SQLAlchemy, but not the slightest idea of how to generate the insert and update data. Any suggestion

I get this error; dem would like an orientation thank you**

![Selección_013](https://user-images.githubusercontent.com/7560024/137216567-536d2818-614e-4ca3-9793-da662b560eda.png)
